controllers/cloud.redhat.com/jobinvocation_controller.go

```
@@ -0,0 +1,240 @@
+	}
+
+	j.Spec.Template.Spec.ImagePullSecrets = []core.LocalObjectReference{
+		{Name: "quay-cloudservices-pull"},
+	}
+
+	envvar := pod.Env
+	envvar = append(envvar, core.EnvVar{Name: "ACG_CONFIG", Value: "/cdapp/cdappconfig.json"})
+
+	var livenessProbe core.Probe
+	var readinessProbe core.Probe
+
+	if pod.LivenessProbe != nil {
+		livenessProbe = *pod.LivenessProbe
+	} else {
+		livenessProbe = core.Probe{}
+	}
+	if pod.ReadinessProbe != nil {
+		readinessProbe = *pod.ReadinessProbe
+	} else {
+		readinessProbe = core.Probe{}
+	}
+
+	c := core.Container{
+		Name:         nn.Name,
+		Image:        pod.Image,
+		Command:      pod.Command,
+		Args:         pod.Args,
+		Env:          envvar,
+		Resources:    maker.ProcessResources(&pod, env),
+		VolumeMounts: pod.VolumeMounts,
+		Ports: []core.ContainerPort{{
+			Name:          "metrics",
+			ContainerPort: env.Spec.Providers.Metrics.Port,
+		}},
+		ImagePullPolicy: core.PullIfNotPresent,
+	}
+
+	if (core.Probe{}) != livenessProbe {
+		c.LivenessProbe = &livenessProbe
+	}
+	if (core.Probe{}) != readinessProbe {
+		c.ReadinessProbe = &readinessProbe
+	}
+
+	c.VolumeMounts = append(c.VolumeMounts, core.VolumeMount{
+		Name:      "config-secret",
+		MountPath: "/cdapp/",
+	})
+
+	j.Spec.Template.Spec.Containers = []core.Container{c}
+
+	j.Spec.Template.Spec.InitContainers = maker.ProcessInitContainers(nn, &c, pod.InitContainers)
+
+	j.Spec.Template.Spec.Volumes = pod.Volumes
+	j.Spec.Template.Spec.Volumes = append(j.Spec.Template.Spec.Volumes, core.Volume{
+		Name: "config-secret",
+		VolumeSource: core.VolumeSource{
+			Secret: &core.SecretVolumeSource{
+				SecretName: app.ObjectMeta.Name,
+			},
+		},
+	})
+
+	maker.ApplyPodAntiAffinity(&j.Spec.Template)
+
+	annotations := make(map[string]string)
+	//annotations["configHash"] = hash
+	j.Spec.Template.SetAnnotations(annotations)
+
+	return nil
+}
+
+func matchAndReturnJob(jobName string, app *crd.ClowdApp) (crd.Job, error) {
```

Let's go with the more standard way of using k8s resources in "getters", don't return the object itself, as that will likely result in a copy, let's create a blank object in the calling function, pass the pointer to it, and then we only need to return the error which can be nil.